packages/pn-commons/src/utility/paymentCaching.utility.ts

```diff
@@ -0,0 +1,133 @@
+import _ from 'lodash';
+
+import { PaymentCache, PaymentDetails, PaymentStatus } from '../models';
+import { paymentCacheSchema } from '../models/PaymentCache';
+
+export const PAYMENT_CACHE_KEY = 'payments';
+
+export const getPaymentCache = (): PaymentCache | null => {
+  const paymentCache = sessionStorage.getItem(PAYMENT_CACHE_KEY);
+  if (!paymentCache) {
+    return null;
+  }
+
+  try {
+    paymentCacheSchema.validateSync(JSON.parse(paymentCache), {
+      stripUnknown: false,
+    });
+  } catch (error) {
+    if (process.env.NODE_ENV === 'development') {
+      console.warn(error);
+    }
+    return null;
+  }
+
+  return paymentCache ? (JSON.parse(paymentCache) as PaymentCache) : null;
+};
+
+export const setPaymentCache = (updatedObj: Partial<PaymentCache>): void => {
+  const paymentCache = getPaymentCache();
+
+  const newPaymentCache = {
+    ...paymentCache,
+    ...updatedObj,
+  } as PaymentCache;
+
+  sessionStorage.setItem(PAYMENT_CACHE_KEY, JSON.stringify(newPaymentCache));
+};
+
+export const setPaymentsInCache = (payments: Array<PaymentDetails>): void => {
+  const paymentCache = getPaymentCache()?.payments;
+
+  if (paymentCache) {
+    const newPaymentCache = _.uniqWith(
+      [...payments, ...paymentCache],
+      (a, b) =>
+        a.pagoPa?.noticeCode === b.pagoPa?.noticeCode &&
+        a.pagoPa?.creditorTaxId === b.pagoPa?.creditorTaxId
+    );
+
+    setPaymentCache({ payments: newPaymentCache });
+  } else {
+    setPaymentCache({ payments });
+  }
+};
+
+export const deletePropertiesInPaymentCache = (properties: Array<keyof PaymentCache>): void => {
+  const paymentCache = getPaymentCache();
+
+  if (paymentCache) {
+    properties.forEach((property) => {
+      // eslint-disable-next-line functional/immutable-data
+      delete paymentCache[property];
+    });
+
+    sessionStorage.setItem(PAYMENT_CACHE_KEY, JSON.stringify(paymentCache));
+  }
+};
+
+export const clearPaymentCache = (): void => {
+  sessionStorage.removeItem(PAYMENT_CACHE_KEY);
+};
+
+// Timestamp is in ISO format
+export const isTimestampWithin2Minutes = (timestamp1: string, timestamp2: string): boolean => {
```

Comment:
isTimestampWithin2Minutes and checkIunAndTimestamp are part of the validation process and they should be called every time we call getPaymentCache.
So i would procede in this way:

1. Create an internal function named validateCache
2. The above function call the schema validation, the isTimestampWithin2Minutes and the checkIunAndTimestamp methods
3. the json.parse must be moved into validation function and used for the next checks
4. isTimestampWithin2Minutes returns true or throw an error (this will be handled by the try catch in getPaymentCache
5. checkIunAndTimestamp becomes checkIun and accepts request iun and cached iun as parameters. it returns true or throw an error (this will be handled by the try catch in getPaymentCache 
6. move the clearPaymentCache from checkIunAndTimestamp to the getPaymentCache in the catch block
7. the setPaymentCache into checkIunAndTimestamp should be useless
8. rework the tests with the goal to cover all cases (also error cases)